client/src/main/java/org/eclipse/hono/client/impl/AmqpAdapterClientCommandResponseSender.java

```diff
@@ -0,0 +1,131 @@
+import org.eclipse.hono.client.CommandResponseSender;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.device.amqp.CommandResponder;
+import org.eclipse.hono.client.device.amqp.TraceableCommandResponder;
+import org.eclipse.hono.util.MessageHelper;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonHelper;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonSender;
+
+/**
+ * A Vertx-Proton based client for sending response messages to commands to Hono's AMQP adapter.
+ */
+public class AmqpAdapterClientCommandResponseSender extends CommandResponseSenderImpl
+        implements CommandResponder, TraceableCommandResponder {
+
+    AmqpAdapterClientCommandResponseSender(final HonoConnection connection, final ProtonSender sender,
+            final String tenantId) {
+        super(connection, sender, tenantId, null);
+    }
+
+    /**
+     * Creates a new sender to send responses for commands back to the business application.
+     *
+     * @param con The connection to the Hono server.
+     * @param tenantId The tenant that the events will be published for.
+     * @param remoteCloseHook The handler to invoke when the link is closed by the peer (may be {@code null}). The
+     *            sender's target address is provided as an argument to the handler.
+     * @return A future indicating the outcome.
+     * @throws NullPointerException if con or tenantId is {@code null}.
+     */
+    public static Future<CommandResponder> createWithAnonymousLinkAddress(
+            final HonoConnection con,
+            final String tenantId,
+            final Handler<String> remoteCloseHook) {
+
+        Objects.requireNonNull(con);
+        Objects.requireNonNull(tenantId);
+
+        return con.createSender(null, ProtonQoS.AT_LEAST_ONCE, remoteCloseHook)
+                .map(sender -> new AmqpAdapterClientCommandResponseSender(con, sender, tenantId));
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendCommandResponse(final String deviceId, final String targetAddress,
+            final String correlationId, final int status, final byte[] payload, final String contentType,
+            final Map<String, ?> properties) {
+        return sendCommandResponse(deviceId, targetAddress, correlationId, status, payload, contentType,
+                properties, null);
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendCommandResponse(final String deviceId, final String targetAddress,
+            final String correlationId, final int status, final byte[] payload,
+            final String contentType, final Map<String, ?> properties, final SpanContext context) {
+
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(targetAddress);
+        Objects.requireNonNull(correlationId);
+
+        final Message message = ProtonHelper.message();
```

Comment:
I think we should also set a message id here.

Reply:
This is done further down in `AbstractSender#sendMessageAndWaitForOutcome(Message, Span)`.

Reply:
Ok, right, I've overlooked that.